src/common/options.cc

```diff
@@ -3036,6 +3036,26 @@ std::vector<Option> get_global_options() {
     .set_description("mclock anticipation timeout in seconds")
     .set_long_description("the amount of time that mclock waits until the unused resource is forfeited"),
 
+    Option("osd_mclock_max_client_iops", Option::TYPE_UINT, Option::LEVEL_BASIC)
```

Comment:
> This makes it tough to tell what sorts of parameters are actually being used by the system - if someone wanted to tune it, they wouldn't know what the current profile was actually using.
> 
> If you have each profile change the settings for the running daemon (e.g. using set_val() on osd_mclock_scheduler_client_res and the other 8 settings when the mclock profile is set) then the code becomes much simpler and it's trivial to inspect what these settings actually are via the usual config show or osd tell commands.
> 
> I don't think the class hierarchy is necessary here - there's no behavior to encapsulate that I can see - it's 3 different ways to set 9 values - and they're static as long as the profile is the same. It seems like 3 functions (or a switch statement) for initializing those 9 settings would be easier to read with the same functionality.

I removed the class hierarchy as suggested and introduced 3 functions that set the profile specific mclock options. Setting Ceph options for the profiles is still WIP but in the meantime please do review this change.

Reply:
> having two parameters described as limits here is kind of confusing, since they're each only used as limits within certain profiles, and the values aren't directly used as limits.
> 
> It seems more natural for the user to describe the total IOPS of the OSD, and for each of the profiles to split that up among client/recovery/best_effort.
> 
> Having a 2nd recovery iops parameter is pretty similar to using a custom profile and manually changing the recovery limit - it seems like an unnecessary degree of freedom.

As per our discussion, having a single option to describe the overall osd capacity would require estimates of IOPS for different block sizes that may be used by the clients. A base starting point of IOPS measured at 4K would be necessary as an input to the cost function. The estimates for higher block sizes can be provided by a cost function for the near term. For the longer term, the benchmark tool would be able to provide more accurate values. 

For the estimates to be set within dmclock server, the block sizes being used by dmclock clients would need to be notified to the mClockScheduler which would in-turn set the IOPS limits and reservations within the dmclock server. This mechanism is in addition to the existing mechanism to set the mclock parameters during startup.